pkg/kubelet/kubelet_pods.go

```diff
@@ -1447,7 +1493,11 @@ func getPhase(pod *v1.Pod, info []v1.ContainerStatus, podIsTerminal bool) v1.Pod
 	}
 
 	switch {
-	case pendingInitialization > 0:
+	case pendingInitialization > 0 &&
+		// This is needed to handle the case where the pod has been initialized but
+		// the restartable init containers are restarting and the pod should not be
+		// placed back into v1.PodPending since the regular containers have run.
+		!kubecontainer.HasAnyRegularContainerStarted(&spec, info):
```

Comment:
As I am reading this I am slightly uncomfortable with checking pod status container status for this.  I *think* it is safe for now, but the source of truth for whether a container has transitioned phases is incompletely modelled in the kubelet loop.  I think in the future we may want to have the pod state machine indicate a one way transition from initializing to initialized through a higher mechanism (like podIsTerminal) which would then be passed down here.

For now it's ok, but we may want to add an assertion somewhere else (will look).

Reply:
xref: https://github.com/kubernetes/kubernetes/issues/126650